analyzers/src/SonarAnalyzer.Common/Helpers/MethodParameterLookupBase.cs

```diff
@@ -18,115 +18,139 @@
+            parameter = methodSymbol.Parameters.FirstOrDefault(symbol => symbol.Name == nameColonArgumentIdentifier.ValueText);
+            return parameter != null;
+        }
 
-            if (!argumentList.HasValue
-                || !argumentList.Value.Contains(arg)
-                || MethodSymbol == null
-                || MethodSymbol.IsVararg)
-            {
-                return false;
-            }
+        var index = argumentList.Value.IndexOf(arg);
+        if (index >= methodSymbol.Parameters.Length)
+        {
+            var lastParameter = methodSymbol.Parameters.Last();
+            parameter = lastParameter.IsParams ? lastParameter : null;
+            return parameter != null;
+        }
+        parameter = methodSymbol.Parameters[index];
+        return true;
+    }
 
-            if (GetNameColonArgumentIdentifier(arg) is { } nameColonArgumentIdentifier)
-            {
-                parameter = MethodSymbol.Parameters.FirstOrDefault(symbol => symbol.Name == nameColonArgumentIdentifier.ValueText);
-                return parameter != null;
-            }
+    /// <summary>
+    /// Method returns array of argument syntaxes that represents all syntaxes passed to the parameter.
+    ///
+    /// There could be multiple syntaxes for ParamArray/params.
+    /// There could be zero or one result for optional parameters.
+    /// There will be single result for normal parameters.
+    /// </summary>
+    public bool TryGetSyntax(IParameterSymbol parameter, out ImmutableArray<SyntaxNode> expressions) =>
+        TryGetSyntax(parameter.Name, out expressions);
 
-            var index = argumentList.Value.IndexOf(arg);
-            if (index >= MethodSymbol.Parameters.Length)
+    /// <summary>
+    /// Method returns array of argument syntaxes that represents all syntaxes passed to the parameter.
+    ///
+    /// There could be multiple syntaxes for ParamArray/params.
+    /// There could be zero or one result for optional parameters.
+    /// There will be single result for normal parameters.
+    public bool TryGetSyntax(string parameterName, out ImmutableArray<SyntaxNode> expressions)
+    {
+        expressions = Enumerable.Empty<SyntaxNode>().ToImmutableArray();
+        if (MethodSymbol is not null)
+        {
+            expressions = GetAllArgumentParameterMappings().Where(x => x.Symbol.Name == parameterName).Select(x => Expression(x.Node)).ToImmutableArray();
+        }
+        else if (MethodSymbolInfo is not null)
+        {
+            var candidateArgumentLists = MethodSymbolInfo.Value.CandidateSymbols.OfType<IMethodSymbol>()
+                .Select(x => GetAllArgumentParameterMappings(x).Where(x => x.Symbol.Name == parameterName).Select(x => Expression(x.Node)).ToImmutableArray()).ToImmutableArray();
+            if (candidateArgumentLists.Any() && candidateArgumentLists.Skip(1).All(x => x.SequenceEqual(candidateArgumentLists[0])))
```

Comment:
This is hard to understand. I would go with speaking parameter names to improve readability. Maybe like so:
```suggestion
            if (candidateArgumentLists.Any()
                && candidateArgumentLists[0] is var firstCandidateMatchingArguments
                && candidateArgumentLists.Skip(1).All(otherCandidateMatchingArguments => otherCandidateMatchingArguments.SequenceEqual(firstCandidateMatchingArguments)))
```

Reply:
I opted for extracting a method instead.